File: Editorial/0164_Maximum_Gap/Go/2_Radix_Sort.go
func maximumGap(nums []int) int {
    if nums == nil || len(nums) < 2 {
        return 0
    }

    maxVal := nums[0]
    for _, num := range nums {
        if num > maxVal {
            maxVal = num
        }
    }
    exp := 1
    radix := 10
    aux := make([]int, len(nums))

    for maxVal/exp > 0 {
        count := make([]int, radix)

        for _, num := range nums {
            count[(num/exp)%10]++
        }

        for i := 1; i < radix; i++ {
            count[i] += count[i-1]
        }

        for i := len(nums) - 1; i >= 0; i-- {
            count[(nums[i]/exp)%10]--
            aux[count[(nums[i]/exp)%10]] = nums[i]
        }

        copy(nums, aux)

        exp *= 10
    }

    maxGap := 0
    for i := 0; i < len(nums)-1; i++ {
        tmp := nums[i+1] - nums[i]
        if tmp > maxGap {
            maxGap = tmp
        }
    }
    return maxGap
}

File: Editorial/0164_Maximum_Gap/Java/2_Radix_Sort.java
class Solution {
    public int maximumGap(int[] nums) {
        if (nums == null || nums.length < 2) {
            return 0;
        }

        int maxVal = Integer.MIN_VALUE;
        for (int num : nums) {
            maxVal = Math.max(maxVal, num);
        }

        int exp = 1;
        int radix = 10;
        int[] aux = new int[nums.length];

        while (maxVal / exp > 0) {
            int[] count = new int[radix];

            for (int num : nums) {
                count[(num / exp) % 10]++;
            }

            for (int i = 1; i < count.length; i++) {
                count[i] += count[i - 1];
            }

            for (int i = nums.length - 1; i >= 0; i--) {
                aux[--count[(nums[i] / exp) % 10]] = nums[i];
            }

            System.arraycopy(aux, 0, nums, 0, nums.length);

            exp *= 10;
        }
        int maxGap = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            maxGap = Math.max(nums[i + 1] - nums[i], maxGap);
        }
        return maxGap;
    }
}
